### ISSUE TITLE
A lint to detect when a type implements all three of `Hash`, `Borrow<str>` and `Borrow<[u8]>`

### ISSUE THREAD

**Issue by thomcc** [rust-lang/rust-clippy#11710]
### What it does

Detects when a type which implements `Hash` also implements both `Borrow<str>` and `Borrow<[u8]>`.

### Advantage

It's impossible to implement both of these correctly, as `str` and `[u8]` hash differently in incompatible ways, and implementing `Borrow<T>` for your type guaratees that you hash equivalent to `T`.

This is worth adding because this impl is tempting to add, as the distinction between `Borrow` and `AsRef` is very subtle, and the `AsRef` impl for both is correct (as is implementing both for types which do not implement `Hash`).

See https://github.com/BurntSushi/bstr/issues/168 for some discussion (As this would not *quite* have caught the problem in that issue, it's possible my formulation of it is too narrow. For the most part, I am chucking this lint proposal over the wall, and am not tied to the specifics here).

### Drawbacks

It's rather niche.

### Example

```rust
impl Hash for MyType {
    // ...
}

impl Borrow<str> for MyType {
    // ...
}

impl Borrow<[u8]> for MyType {
    // ...
}
```

is wrong, either remove any one of those impls to be correct.

**Comment by partiallytyped** [comment]
@rustbot claim

This one looks interesting

**Comment by partiallytyped** [comment]
Maybe there should be another pedantic lint warning against implementing `Borrow<T: Hash> and Hash` unless the Hash impl calls borrow internally. Can this be identified?

Possibly another for implementation of `Borrow<T:Hash>` for multiple T warning that they should be identical.

This is a can of worms :D
